Orange/canvas/utils/overlay.py

```diff
@@ -498,84 +544,105 @@ def button(self, standardButton):
-            # each canvas requires its own instance of each notification
-            cloned = notif.clone()
-            cloned.setParent(overlay)
-            cloned.clicked.connect(NotificationOverlay.nextNotification)
-            overlay.addWidget(cloned)
+    @Slot(Notification)
+    def addNotification(self, notif: Notification):
+        notifWidget = NotificationWidget.fromNotification(notif, parent=self)
 
-    @staticmethod
-    def nextNotification():
-        NotificationOverlay.notifQueue.pop(0)
+        notifWidget.clicked.connect(lambda b:
+                                    notif.clicked.emit(notifWidget.buttonRole(b)))
+        notifWidget.accepted.connect(notif.accepted)
+        notifWidget.rejected.connect(notif.rejected)
+        notifWidget.dismissed.connect(notif.dismissed)
 
-        overlays = NotificationOverlay.overlayInstances
-        for overlay in overlays:
-            overlay.nextWidget()
+        self._addWidget(notifWidget)
 
-    def addWidget(self, widget):
+    @Slot()
+    def nextWidget(self):
+        """
+        Removes first widget from the stack.
+        """
+        if not self._widgets:
+            log.error("Received next notification signal while no notification is displayed")
+            return
+        widget: NotificationWidget = self._widgets.pop(0)
+        self.layout().removeWidget(widget)
+        widget.close()
+
+    def _addWidget(self, widget):
         """
         Append the widget to the stack.
         """
         self._widgets.append(widget)
         self.layout().addWidget(widget)
 
-    def nextWidget(self):
-        """
-        Removes first widget from the stack.
 
-        .. note:: The widget is hidden but is not deleted.
+class NotificationServer(QObject, metaclass=Singleton):
+    """
+    NotificationServer is a singleton – call the class method
+    `NotificationServer.registerNotification(notif: Notification)` to display a notification.
+    """
+    # emits when new notification is registered
+    newNotification = Signal(Notification)
+
+    # emits when a notification is responded to
+    nextNotification = Signal()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # list of queued Notification objects, to populate newly generated canvases
+        self._notificationQueue = []
+
+        self.nextNotification.connect(self._nextNotification)
 
+    @classmethod
+    def registerNotification(cls, notif: Notification):
         """
-        widget = self._widgets[0]
-        self._widgets.pop(0)
-        self.layout().removeWidget(widget)
+        :type notif: Notification
 
-    def currentWidget(self):
+        After instantiating a Notification, use this method to send it.
+        Queues notification in all canvas instances (shows it if no other notifications present).
         """
-        Return the currently displayed widget.
+        # get singleton instance
+        self = cls()
```

Comment:
I seem to have missed https://github.com/biolab/orange3/pull/3933#discussion_r304329001

Reply:
@ales-erjavec, I think that we could still avoid singletons. If an instance is created in main and simply attached to something (module?) that is accessible anywhere, it would still address my concern.

Reply:
Like so?